Post-Processing slows down to a crawl and fails with large globs

@jsotobroad -- take it away

In a scatter of 20500 shards, we ran a task that basically took in one file input and output a glob of files.  We first tried this with a glob where we expected ~900 files to be output and no memory issues were found and everything went relatively smoothly.  Because of some outside factors we decided to change this task to instead output ~3000 files in the glob.  After about 13000 tasks were processed(Sucess -> Done) we started seeing some slow down that coincided with errors in the logs like the following:
```
2016-08-03 03:34:04,971 cromwell-system-akka.actor.default-dispatcher-51 WARN  - Caught exception, retrying: Remote host closed connection during handshake
javax.net.ssl.SSLHandshakeException: Remote host closed connection during handshake
        at sun.security.ssl.SSLSocketImpl.readRecord(SSLSocketImpl.java:992) ~[na:1.8.0_72]
        at sun.security.ssl.SSLSocketImpl.performInitialHandshake(SSLSocketImpl.java:1375) ~[na:1.8.0_72]
        at sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:1403) ~[na:1.8.0_72]
        at sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:1387) ~[na:1.8.0_72]
        at sun.net.www.protocol.https.HttpsClient.afterConnect(HttpsClient.java:559) ~[na:1.8.0_72]
        at sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.connect(AbstractDelegateHttpsURLConnection.java:185) ~[na:1.8.0_72]
        at sun.net.www.protocol.https.HttpsURLConnectionImpl.connect(HttpsURLConnectionImpl.java:153) ~[na:1.8.0_72]
        at com.google.api.client.http.javanet.NetHttpRequest.execute(NetHttpRequest.java:93) ~[cromwell.jar:0.19]
        at com.google.api.client.http.HttpRequest.execute(HttpRequest.java:972) ~[cromwell.jar:0.19]
        at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:419) ~[cromwell.jar:0.19]
        at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:352) ~[cromwell.jar:0.19]
        at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.execute(AbstractGoogleClientRequest.java:469) ~[cromwell.jar:0.19]
        at cromwell.engine.backend.io.filesystem.gcs.GcsFileSystemProvider$$anonfun$3.apply(GcsFileSystemProvider.scala:242) ~[cromwell.jar:0.19]
        at cromwell.engine.backend.io.filesystem.gcs.GcsFileSystemProvider$$anonfun$3.apply(GcsFileSystemProvider.scala:242) ~[cromwell.jar:0.19]
        at scala.util.Try$.apply(Try.scala:192) ~[cromwell.jar:0.19]
        at cromwell.engine.backend.io.filesystem.gcs.GcsFileSystemProvider$.withRetry(GcsFileSystemProvider.scala:48) ~[cromwell.jar:0.19]
        at cromwell.engine.backend.io.filesystem.gcs.GcsFileSystemProvider.requestListPage$1(GcsFileSystemProvider.scala:242) ~[cromwell.jar:0.19]
        at cromwell.engine.backend.io.filesystem.gcs.GcsFileSystemProvider.remainingObjects$1(GcsFileSystemProvider.scala:250) ~[cromwell.jar:0.19]
        at cromwell.engine.backend.io.filesystem.gcs.GcsFileSystemProvider.list(GcsFileSystemProvider.scala:256) ~[cromwell.jar:0.19]
        at cromwell.engine.backend.io.filesystem.gcs.GcsFileSystemProvider.newDirectoryStream(GcsFileSystemProvider.scala:265) ~[cromwell.jar:0.19]
        at java.nio.file.Files.newDirectoryStream(Files.java:457) ~[na:1.8.0_72]
        at better.files.File.list(File.scala:200) ~[cromwell.jar:0.19]
        at cromwell.engine.backend.jes.JesWorkflowEngineFunctions.glob(JesEngineFunctions.scala:15) ~[cromwell.jar:0.19]
        at cromwell.engine.WdlStandardLibraryImpl$$anonfun$glob$1.apply(WdlStandardLibraryImpl.scala:171) ~[cromwell.jar:0.19]
        at cromwell.engine.WdlStandardLibraryImpl$$anonfun$glob$1.apply(WdlStandardLibraryImpl.scala:169) ~[cromwell.jar:0.19]
        at scala.util.Success$$anonfun$map$1.apply(Try.scala:237) ~[cromwell.jar:0.19]
        at scala.util.Try$.apply(Try.scala:192) ~[cromwell.jar:0.19]
        at scala.util.Success.map(Try.scala:237) ~[cromwell.jar:0.19]
        at cromwell.engine.WdlStandardLibraryImpl$class.glob(WdlStandardLibraryImpl.scala:169) ~[cromwell.jar:0.19]
        at cromwell.engine.backend.WorkflowEngineFunctions.glob(EngineFunctions.scala:15) ~[cromwell.jar:0.19]
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.8.0_72]
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:1.8.0_72]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_72]
        at java.lang.reflect.Method.invoke(Method.java:498) ~[na:1.8.0_72]
        at wdl4s.expression.WdlFunctions$$anonfun$getFunction$1.apply(WdlFunctions.scala:21) ~[cromwell.jar:0.19]
        at wdl4s.expression.WdlFunctions$$anonfun$getFunction$1.apply(WdlFunctions.scala:21) ~[cromwell.jar:0.19]
        at wdl4s.expression.ValueEvaluator.evaluate(ValueEvaluator.scala:126) ~[cromwell.jar:0.19]
        at wdl4s.WdlExpression$.evaluate(WdlExpression.scala:79) ~[cromwell.jar:0.19]
        at wdl4s.WdlExpression.evaluate(WdlExpression.scala:192) ~[cromwell.jar:0.19]
        at cromwell.engine.backend.jes.JesBackend.cromwell$engine$backend$jes$JesBackend$$outputLookup(JesBackend.scala:695) ~[cromwell.jar:0.19]
        at cromwell.engine.backend.jes.JesBackend$$anonfun$getOutputFoldingFunction$1.apply(JesBackend.scala:675) ~[cromwell.jar:0.19]
        at cromwell.engine.backend.jes.JesBackend$$anonfun$getOutputFoldingFunction$1.apply(JesBackend.scala:674) ~[cromwell.jar:0.19]
        at scala.collection.LinearSeqOptimized$class.foldLeft(LinearSeqOptimized.scala:124) ~[cromwell.jar:0.19]
        at scala.collection.immutable.List.foldLeft(List.scala:84) ~[cromwell.jar:0.19]
        at cromwell.engine.backend.jes.JesBackend.postProcess(JesBackend.scala:665) ~[cromwell.jar:0.19]
        at cromwell.engine.backend.jes.JesBackend$$anonfun$executionResult$1.apply(JesBackend.scala:705) ~[cromwell.jar:0.19]
        at cromwell.engine.backend.jes.JesBackend$$anonfun$executionResult$1.apply(JesBackend.scala:705) ~[cromwell.jar:0.19]
        at cromwell.engine.backend.jes.JesBackend$$anonfun$executionResult$1.apply(JesBackend.scala:700) ~[cromwell.jar:0.19]
        at scala.concurrent.impl.Future$PromiseCompletingRunnable.liftedTree1$1(Future.scala:24) ~[cromwell.jar:0.19]
        at scala.concurrent.impl.Future$PromiseCompletingRunnable.run(Future.scala:24) ~[cromwell.jar:0.19]
        at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:40) ~[cromwell.jar:0.19]
        at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(AbstractDispatcher.scala:397) [cromwell.jar:0.19]
        at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260) [cromwell.jar:0.19]
        at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339) [cromwell.jar:0.19]
        at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979) [cromwell.jar:0.19]
        at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107) [cromwell.jar:0.19]
Caused by: java.io.EOFException: SSL peer shut down incorrectly
        at sun.security.ssl.InputRecord.read(InputRecord.java:505) ~[na:1.8.0_72]
        at sun.security.ssl.SSLSocketImpl.readRecord(SSLSocketImpl.java:973) ~[na:1.8.0_72]
        ... 54 common frames omitted
```
and
```
2016-08-03 03:33:06,985 cromwell-system-akka.actor.default-dispatcher-3 WARN  - Caught exception, retrying: Broken pipe
java.net.SocketException: Broken pipe
        at java.net.SocketOutputStream.socketWrite0(Native Method) ~[na:1.8.0_72]
        at java.net.SocketOutputStream.socketWrite(SocketOutputStream.java:109) ~[na:1.8.0_72]
        at java.net.SocketOutputStream.write(SocketOutputStream.java:153) ~[na:1.8.0_72]
        at sun.security.ssl.OutputRecord.writeBuffer(OutputRecord.java:431) ~[na:1.8.0_72]
        at sun.security.ssl.OutputRecord.write(OutputRecord.java:417) ~[na:1.8.0_72]
        at sun.security.ssl.SSLSocketImpl.writeRecordInternal(SSLSocketImpl.java:876) ~[na:1.8.0_72]
        at sun.security.ssl.SSLSocketImpl.writeRecord(SSLSocketImpl.java:847) ~[na:1.8.0_72]
        at sun.security.ssl.SSLSocketImpl.writeRecord(SSLSocketImpl.java:717) ~[na:1.8.0_72]
        at sun.security.ssl.Handshaker.sendChangeCipherSpec(Handshaker.java:1077) ~[na:1.8.0_72]
        at sun.security.ssl.ClientHandshaker.sendChangeCipherAndFinish(ClientHandshaker.java:1222) ~[na:1.8.0_72]
        at sun.security.ssl.ClientHandshaker.serverHelloDone(ClientHandshaker.java:1134) ~[na:1.8.0_72]
        at sun.security.ssl.ClientHandshaker.processMessage(ClientHandshaker.java:348) ~[na:1.8.0_72]
        at sun.security.ssl.Handshaker.processLoop(Handshaker.java:979) ~[na:1.8.0_72]
        at sun.security.ssl.Handshaker.process_record(Handshaker.java:914) ~[na:1.8.0_72]
        at sun.security.ssl.SSLSocketImpl.readRecord(SSLSocketImpl.java:1062) ~[na:1.8.0_72]
        at sun.security.ssl.SSLSocketImpl.performInitialHandshake(SSLSocketImpl.java:1375) ~[na:1.8.0_72]
        at sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:1403) ~[na:1.8.0_72]
        at sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:1387) ~[na:1.8.0_72]
        at sun.net.www.protocol.https.HttpsClient.afterConnect(HttpsClient.java:559) ~[na:1.8.0_72]
        at sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.connect(AbstractDelegateHttpsURLConnection.java:185) ~[na:1.8.0_72]
        at sun.net.www.protocol.https.HttpsURLConnectionImpl.connect(HttpsURLConnectionImpl.java:153) ~[na:1.8.0_72]
        at com.google.api.client.http.javanet.NetHttpRequest.execute(NetHttpRequest.java:93) ~[cromwell.jar:0.19]
        at com.google.api.client.http.HttpRequest.execute(HttpRequest.java:972) ~[cromwell.jar:0.19]
        at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:419) ~[cromwell.jar:0.19]
        at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:352) ~[cromwell.jar:0.19]
        at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.execute(AbstractGoogleClientRequest.java:469) ~[cromwell.jar:0.19]
        at cromwell.engine.backend.jes.Run.status(Run.scala:143) ~[cromwell.jar:0.19]
        at cromwell.engine.backend.jes.Run.checkStatus(Run.scala:156) ~[cromwell.jar:0.19]
        at cromwell.engine.backend.jes.JesBackend$$anonfun$poll$1$$anonfun$42.apply(JesBackend.scala:933) ~[cromwell.jar:0.19]
        at cromwell.engine.backend.jes.JesBackend$$anonfun$poll$1$$anonfun$42.apply(JesBackend.scala:933) ~[cromwell.jar:0.19]
        at scala.util.Try$.apply(Try.scala:192) ~[cromwell.jar:0.19]
        at cromwell.engine.backend.jes.JesBackend$$anonfun$poll$1.apply(JesBackend.scala:933) [cromwell.jar:0.19]
        at cromwell.engine.backend.jes.JesBackend$$anonfun$poll$1.apply(JesBackend.scala:927) [cromwell.jar:0.19]
        at scala.concurrent.impl.Future$PromiseCompletingRunnable.liftedTree1$1(Future.scala:24) [cromwell.jar:0.19]
        at scala.concurrent.impl.Future$PromiseCompletingRunnable.run(Future.scala:24) [cromwell.jar:0.19]
        at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:40) [cromwell.jar:0.19]
        at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(AbstractDispatcher.scala:397) [cromwell.jar:0.19]
        at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260) [cromwell.jar:0.19]
        at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339) [cromwell.jar:0.19]
        at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979) [cromwell.jar:0.19]
        at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107) [cromwell.jar:0.19]
```
We suspect that of all the retries from the errors above plus the new succeeded tasks that now needed post processing created a backlog that may have cause Cromwell to become sluggish and eventually run out of memory.  It may have also been the opposite direction where Cromwell becoming sluggish caused these errors to start popping up.

@dshiga any additions to explain what we saw when you launched the 20k callset with 3000 intervals?

--- a/supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/JesAsyncBackendJobExecutionActor.scala
+++ b/supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/JesAsyncBackendJobExecutionActor.scala
@@ -65,7 +65,7 @@ object JesAsyncBackendJobExecutionActor {
     * ask them for results.
     */
   case class JesPendingExecutionHandle(jobDescriptor: BackendJobDescriptor,
-                                       jesOutputs: Seq[JesFileOutput],
+                                       jesOutputs: Set[JesFileOutput],
                                        run: Run,
                                        previousStatus: Option[RunStatus]) extends ExecutionHandle {
     override val isDone = false
@@ -126,8 +126,6 @@ class JesAsyncBackendJobExecutionActor(override val jobDescriptor: BackendJobDes
 
   override def receive: Receive = pollingActorClientReceive orElse jesReceiveBehavior orElse super.receive
 
-  private def globOutputPath(glob: String) = callRootPath.resolve(s"glob-${glob.md5Sum}/")
-
   private def gcsAuthParameter: Option[JesInput] = {
     if (jesAttributes.auths.gcs.requiresAuthFile || dockerConfiguration.isDefined)
       Option(JesLiteralInput(ExtraConfigParamName, jesCallPaths.gcsAuthFilePath.toUri.toString))
@@ -163,15 +161,14 @@ class JesAsyncBackendJobExecutionActor(override val jobDescriptor: BackendJobDes
     */
   private def relativeLocalizationPath(file: WdlFile): WdlFile = {
     getPath(file.value) match {
-      case Success(path) => {
+      case Success(path) =>
         val value: WdlSource = path.toUri.getHost + path.toUri.getPath
         WdlFile(value, file.isGlob)
-      }
       case _ => file
     }
   }
 
-  private[jes] def generateJesInputs(jobDescriptor: BackendJobDescriptor): Iterable[JesInput] = {
+  private[jes] def generateJesInputs(jobDescriptor: BackendJobDescriptor): Set[JesInput] = {
 
     val writeFunctionFiles = call.task.evaluateFilesFromCommand(jobDescriptor.fullyQualifiedInputs, callEngineFunctions) map {
       case (expression, file) =>  expression.toWdlString.md5SumShort -> Seq(file)
@@ -180,9 +177,11 @@ class JesAsyncBackendJobExecutionActor(override val jobDescriptor: BackendJobDes
     /** Collect all WdlFiles from inputs to the call */
     val callInputFiles: Map[FullyQualifiedName, Seq[WdlFile]] = jobDescriptor.fullyQualifiedInputs mapValues { _.collectAsSeq { case w: WdlFile => w } }
 
-    (callInputFiles ++ writeFunctionFiles) flatMap {
+    val inputs = (callInputFiles ++ writeFunctionFiles) flatMap {
       case (name, files) => jesInputsFromWdlFiles(name, files, files.map(relativeLocalizationPath), jobDescriptor)
     }
+
+    inputs.toSet
   }
 
   /**
@@ -212,18 +211,48 @@ class JesAsyncBackendJobExecutionActor(override val jobDescriptor: BackendJobDes
     if (referenceName.length <= 127) referenceName else referenceName.md5Sum
   }
 
-  private[jes] def generateJesOutputs(jobDescriptor: BackendJobDescriptor): Seq[JesFileOutput] = {
+  private[jes] def findGlobOutputs(jobDescriptor: BackendJobDescriptor): Set[WdlGlobFile] = {
+    val globOutputs = (call.task.findOutputFiles(jobDescriptor.fullyQualifiedInputs, NoFunctions) map relativeLocalizationPath) collect {
+      case glob: WdlGlobFile => glob
+    }
+    globOutputs.distinct.toSet
+  }
+
+  private[jes] def generateJesOutputs(jobDescriptor: BackendJobDescriptor): Set[JesFileOutput] = {
     val wdlFileOutputs = call.task.findOutputFiles(jobDescriptor.fullyQualifiedInputs, NoFunctions) map relativeLocalizationPath
 
-    // Create the mappings. GLOB mappings require special treatment (i.e. stick everything matching the glob in a folder)
-    wdlFileOutputs.distinct map { wdlFile =>
-      val destination = wdlFile match {
-        case WdlSingleFile(filePath) => callRootPath.resolve(filePath.stripPrefix("/")).toUri.toString
-        case WdlGlobFile(filePath) => globOutputPath(filePath).toUri.toString
+    val outputs = wdlFileOutputs.distinct flatMap { wdlFile =>
+      wdlFile match {
+        case singleFile: WdlSingleFile => List(generateJesSingleFileOutputs(singleFile))
+        case globFile: WdlGlobFile => generateJesGlobFileOutputs(globFile)
       }
-      val (relpath, disk) = relativePathAndAttachedDisk(wdlFile.value, runtimeAttributes.disks)
-      JesFileOutput(makeSafeJesReferenceName(wdlFile.value), destination, relpath, disk)
     }
+
+    outputs.toSet
+  }
+
+  private def generateJesSingleFileOutputs(wdlFile: WdlSingleFile): JesFileOutput = {
+    val destination = callRootPath.resolve(wdlFile.value.stripPrefix("/")).toUri.toString
+    val (relpath, disk) = relativePathAndAttachedDisk(wdlFile.value, runtimeAttributes.disks)
+    JesFileOutput(makeSafeJesReferenceName(wdlFile.value), destination, relpath, disk)
+  }
+
+  private def generateJesGlobFileOutputs(wdlFile: WdlGlobFile): List[JesFileOutput] = {
+    val globName = callEngineFunctions.globName(wdlFile.value)
+    val globDirectory = globName + "/"
+    val globListFile = globName + ".list"
+    val gcsGlobDirectoryDestinationPath = callRootPath.resolve(globDirectory).toUri.toString
+    val gcsGlobListFileDestinationPath = callRootPath.resolve(globListFile).toUri.toString
+
+    val (_, globDirectoryDisk) = relativePathAndAttachedDisk(wdlFile.value, runtimeAttributes.disks)
+
+    // We need both the glob directory and the glob list:
+    List(
+      // The glob directory:
+      JesFileOutput(makeSafeJesReferenceName(globDirectory), gcsGlobDirectoryDestinationPath, Paths.get(globDirectory + "*"), globDirectoryDisk),
+      // The glob list file:
+      JesFileOutput(makeSafeJesReferenceName(globListFile), gcsGlobListFileDestinationPath, Paths.get(globListFile), globDirectoryDisk)
+    )
   }
 
   private def instantiateCommand: Try[String] = {
@@ -231,7 +260,7 @@ class JesAsyncBackendJobExecutionActor(override val jobDescriptor: BackendJobDes
     jobDescriptor.call.task.instantiateCommand(backendInputs, callEngineFunctions, valueMapper = gcsPathToLocal)
   }
 
-  private def uploadCommandScript(command: String, withMonitoring: Boolean): Future[Unit] = {
+  private def uploadCommandScript(command: String, withMonitoring: Boolean, globFiles: Set[WdlGlobFile]): Future[Unit] = {
     val monitoring = if (withMonitoring) {
       s"""|touch $JesMonitoringLogFile
           |chmod u+x $JesMonitoringScript
@@ -240,6 +269,22 @@ class JesAsyncBackendJobExecutionActor(override val jobDescriptor: BackendJobDes
 
     val tmpDir = File(JesWorkingDisk.MountPoint)./("tmp").path
     val rcPath = File(JesWorkingDisk.MountPoint)./(returnCodeFilename).path
+
+    def globManipulation(globFile: WdlGlobFile) = {
+
+      val globDir = callEngineFunctions.globName(globFile.value)
+      val (_, disk) = relativePathAndAttachedDisk(globFile.value, runtimeAttributes.disks)
+      val globDirectory = Paths.get(s"${disk.mountPoint.toAbsolutePath}/$globDir/")
+      val globList = Paths.get(s"${disk.mountPoint.toAbsolutePath}/$globDir.list")
+
+      s"""
+        |mkdir $globDirectory
+        |ln ${globFile.value} $globDirectory
+        |ls -1 $globDirectory > $globList
+      """.stripMargin
+    }
+
+    val globManipulations = globFiles.map(globManipulation).mkString("\n")
 
     val fileContent =
       s"""
@@ -250,6 +295,7 @@ class JesAsyncBackendJobExecutionActor(override val jobDescriptor: BackendJobDes
          |(
          |cd ${JesWorkingDisk.MountPoint}
          |$command
+         |$globManipulations
          |)
          |echo $$? > $rcPath
        """.stripMargin.trim
@@ -294,8 +340,8 @@ class JesAsyncBackendJobExecutionActor(override val jobDescriptor: BackendJobDes
   }
 
   protected def runWithJes(command: String,
-                           jesInputs: Seq[JesInput],
-                           jesOutputs: Seq[JesFileOutput],
+                           jesInputs: Set[JesInput],
+                           jesOutputs: Set[JesFileOutput],
                            runIdForResumption: Option[String],
                            withMonitoring: Boolean): Future[ExecutionHandle] = {
 
@@ -304,7 +350,7 @@ class JesAsyncBackendJobExecutionActor(override val jobDescriptor: BackendJobDes
     val jesParameters = standardParameters ++ gcsAuthParameter ++ jesInputs ++ jesOutputs
 
     val jesJobSetup = for {
-      _ <- uploadCommandScript(command, withMonitoring)
+      _ <- uploadCommandScript(command, withMonitoring, findGlobOutputs(jobDescriptor))
       run <- createJesRun(jesParameters, runIdForResumption)
       _ = tellMetadata(Map(CallMetadataKeys.JobId -> run.runId))
     } yield run
@@ -321,8 +367,8 @@ class JesAsyncBackendJobExecutionActor(override val jobDescriptor: BackendJobDes
   }
 
   private def startExecuting(monitoringOutput: Option[JesFileOutput], mode: ExecutionMode): Future[ExecutionHandle] = {
-    val jesInputs: Seq[JesInput] = generateJesInputs(jobDescriptor).toSeq ++ monitoringScript :+ cmdInput
-    val jesOutputs: Seq[JesFileOutput] = generateJesOutputs(jobDescriptor) ++ monitoringOutput
+    val jesInputs: Set[JesInput] = generateJesInputs(jobDescriptor) ++ monitoringScript + cmdInput
+    val jesOutputs: Set[JesFileOutput] = generateJesOutputs(jobDescriptor) ++ monitoringOutput
 
     instantiateCommand match {
       case Success(command) => runWithJes(command, jesInputs, jesOutputs, mode.jobId.collectFirst { case j: JesJobId => j.operationId }, monitoringScript.isDefined)
@@ -419,7 +465,7 @@ class JesAsyncBackendJobExecutionActor(override val jobDescriptor: BackendJobDes
     serviceRegistryActor.putMetadata(jobDescriptor.workflowDescriptor.id, Option(jobDescriptor.key), metadataKeyValues)
   }
 
-  private[jes] def wdlValueToGcsPath(jesOutputs: Seq[JesFileOutput])(value: WdlValue): WdlValue = {
+  private[jes] def wdlValueToGcsPath(jesOutputs: Set[JesFileOutput])(value: WdlValue): WdlValue = {
     def toGcsPath(wdlFile: WdlFile) = jesOutputs collectFirst {
       case o if o.name == makeSafeJesReferenceName(wdlFile.valueString) => WdlFile(o.gcs)
     } getOrElse value

--- a/supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/JesExpressionFunctions.scala
+++ b/supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/JesExpressionFunctions.scala
@@ -10,28 +10,21 @@ import wdl4s.expression.{PureStandardLibraryFunctionsLike, WdlStandardLibraryFun
 import wdl4s.values._
 
 import scala.collection.JavaConverters._
-import scala.language.postfixOps
 import scala.util.{Success, Try}
 
 class JesExpressionFunctions(override val pathBuilders: List[PathBuilder], context: CallContext)
   extends WdlStandardLibraryFunctions with PureStandardLibraryFunctionsLike with ReadLikeFunctions with WriteFunctions {
 
   override def writeTempFile(path: String, prefix: String, suffix: String, content: String): String = super[WriteFunctions].writeTempFile(path, prefix, suffix, content)
-  private def globDirectory(glob: String): String = s"glob-${glob.md5Sum}/"
+  private[jes] def globDirectory(glob: String): String = globName(glob) + "/"
+  private[jes] def globName(glob: String) = s"glob-${glob.md5Sum}"
 
   override def globPath(glob: String): String = context.root.resolve(globDirectory(glob)).toString
 
   override def glob(path: String, pattern: String): Seq[String] = {
-    /* Globbing is not implemented in the gcs-nio implementation yet (specifically PathMatcher) at this time (09/2016)
-     * which is fine here since all files in the globing directory have already be matched against the glob pattern by JES.
-     *
-     * Also, better.file.File can't be used here because it calls toAbsolutePath on the path in File.apply, which adds a leading "/" in the path.
-     * This makes the newDirectoryStream implementation of CloudStorageFileSystemProvider return nothing
-     * because it doesn't call toRealPath for this method before doing I/O and hence does not remove the "/" prefix (bug ?)
-     * See com.google.cloud.storage.contrib.nio.CloudStorageConfiguration#stripPrefixSlash()
-     */
-    val directory = context.root.resolve(s"glob-${pattern.md5Sum}/").toRealPath()
-    Files.list(directory).iterator().asScala filterNot { Files.isDirectory(_) } map { _.toUri.toString } toSeq
+    val name = globName(pattern)
+    val listFile = context.root.resolve(s"$name.list").toRealPath()
+    Files.readAllLines(listFile).asScala map { fileName => context.root.resolve(s"$name/$fileName").toUri.toString }
   }
 
   override def preMapping(str: String): String = if (!GcsPathBuilder.isValidGcsUrl(str)) {

--- a/supportedBackends/jes/src/test/scala/cromwell/backend/impl/jes/JesAsyncBackendJobExecutionActorSpec.scala
+++ b/supportedBackends/jes/src/test/scala/cromwell/backend/impl/jes/JesAsyncBackendJobExecutionActorSpec.scala
@@ -129,7 +129,7 @@ class JesAsyncBackendJobExecutionActorSpec extends TestKitSuite("JesAsyncBackend
 
     // Mock/stub out the bits that would reach out to JES.
     val run = mock[Run]
-    val handle = JesPendingExecutionHandle(jobDescriptor, Seq.empty, run, None)
+    val handle = JesPendingExecutionHandle(jobDescriptor, Set.empty, run, None)
 
     class ExecuteOrRecoverActor extends TestableJesJobExecutionActor(jobDescriptor, promise, jesConfiguration, jesSingletonActor = jesSingletonActor) {
       override def executeOrRecover(mode: ExecutionMode)(implicit ec: ExecutionContext): Future[ExecutionHandle] = Future.successful(handle)
@@ -488,7 +488,7 @@ class JesAsyncBackendJobExecutionActorSpec extends TestKitSuite("JesAsyncBackend
   }
 
   it should "convert local Paths back to corresponding GCS paths in JesOutputs" in {
-    val jesOutputs = Seq(
+    val jesOutputs = Set(
       JesFileOutput("/cromwell_root/path/to/file1", "gs://path/to/file1",
         Paths.get("/cromwell_root/path/to/file1"), workingDisk),
       JesFileOutput("/cromwell_root/path/to/file2", "gs://path/to/file2",

